### PR TITLE
DrivAerML: multi-seed full-eval paper-facing baseline (seed=123)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Paper-facing DrivAerML test results need full evaluation coverage. This is one of three multi-seed full-eval runs (seed=123) to establish a robust test baseline for the 4L/512d champion config. No `--max-eval-batches` truncation.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --seed 123 \
  --wandb_group dm-fulleval-multiseed
```

**Critical:** No `--max-eval-batches`. Full evaluation is mandatory.

## Reporting Requirements
- Best val epoch + `val_primary/surface_rel_l2_pct`
- `test_primary/surface_rel_l2_pct` at best val checkpoint
- Confirm: full eval (no --max-eval-batches used)
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

```bash
# DrivAerML champion (val 3.997% @ ep467)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT casca:

## Results

### DrivAerML full-eval seed=123 (4L/512d champion config)

| Metric | Value |
|---|---|
| **val_primary/surface_rel_l2_pct** | **14.027%** @ epoch 40 |
| **test_primary/surface_rel_l2_pct** | **15.580%** |
| Full eval (no --max-eval-batches) | ✅ Confirmed |
| Epochs completed | 50 (360-min timeout) |
| W&B run ID | `7acr9vfr` |
| W&B group | `dm-fulleval-multiseed` |
| Peak memory | ~6.2 GB (GPU 0) |

### Command used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --seed 123 \
  --wandb-group dm-fulleval-multiseed \
  --agent casca --wandb-name "casca/dm-fulleval-seed123"
```

### Comparison to baseline

- Baseline (truncated eval, --max-eval-batches 200): **3.997%** val @ ep467 (W&B: `bht6h42t`)
- This run (full eval, no --max-eval-batches): **14.027%** val @ ep40
- **Not directly comparable** — full eval ran only 50 epochs vs 467 for the truncated-eval baseline due to ~10x longer per-epoch eval time (~7.2 min/epoch full-eval vs ~0.77 min/epoch truncated)

### What happened

The full-eval run completed 50 epochs in the 360-minute budget. Without `--max-eval-batches`, each epoch takes ~7.2 minutes (eval covers all samples), compared to ~0.77 min/epoch with truncated eval. This limits the training to ~50 epochs vs the 467 epochs achieved in the truncated-eval baseline.

The val metric descended steadily through epochs 1-40 (63.4% → 14.0%), following the characteristic DrivAerML cosine-trough descent pattern. At epoch 45-46, a transient instability spike occurred (val jumped to 148% at ep46) but the model recovered by ep47-49. Best checkpoint was at epoch 40 (14.03%).

The test metric at best checkpoint is **15.58%** — this is the first full-eval test number for the 4L/512d champion config, establishing the paper-facing baseline at seed=123.

Key observation: The full-eval run is severely epoch-limited. The truncated-eval run achieved its best at ep467 after hundreds more epochs of optimization. At only 50 epochs, this seed=123 run was still improving (loss at 0.055, trough envelope still descending). The 14.03% result is an underestimate of the full-eval ceiling.

### Suggested follow-ups

- Compare with the other multi-seed runs (seed=42, seed=456) to assess seed variance under full eval
- Consider whether a longer timeout budget (e.g. 720 min) would allow enough epochs for the full-eval runs to approach the truncated-eval quality
- The instability at ep45-46 suggests `--grad-clip 1.0` could help stabilize longer full-eval training

**Comment by tcapelle** [comment]
ADVISOR: Full eval costs ~7.2 min/epoch — only 50 epochs completed vs 467 needed for convergence. The test metric (15.6%) is from an undertrained model and is not paper-ready. Please use the two-phase approach: (1) train with **--max-eval-batches 200** using seed=123 until convergence at ~450+ epochs, then (2) load the best val checkpoint and run **one final full eval pass without --max-eval-batches** to get the paper-facing test_primary number. Report both val (from fast training) and test (from the single full-eval pass). This is the correct protocol for paper-facing DM results.

**Comment by tcapelle** [comment]
STUDENT casca:

## Results — Two-Phase Protocol (Revision)

### DrivAerML seed=123: truncated-val training + full test eval

| Metric | Value |
|---|---|
| **val_primary/surface_rel_l2_pct** | **11.888%** @ epoch 96 (truncated eval, --max-eval-batches 200) |
| **test_primary/surface_rel_l2_pct** | **12.155%** (full eval, best-val checkpoint — no --max-eval-batches) |
| Full test eval confirmed | ✅ (--full-test-eval flag, test_max_batches_override=0) |
| Epochs completed | 499 (NaN from ep388 onwards) |
| W&B run ID | `vgey7keg` |
| W&B group | `dm-fulleval-multiseed` |
| Runtime | ~383 minutes |

### Command used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --full-test-eval \
  --seed 123 --wandb-group dm-fulleval-multiseed \
  --agent casca --wandb-name "casca/dm-fulleval-seed123-2phase"
```

### Comparison to baseline

| | val_primary | test_primary | Epochs | Eval mode |
|---|---|---|---|---|
| **Baseline (seed=0)** | **3.997%** @ ep467 | 6.244% (older config, truncated) | 467 | truncated |
| **This run (seed=123)** | **11.888%** @ ep96 | **12.155%** | 499 (NaN from 388) | two-phase |

### What happened — training instability

The two-phase protocol worked correctly: truncated val evaluation during training (--max-eval-batches 200), then full test evaluation on the best-val checkpoint at the end (--full-test-eval).

However, seed=123 is dramatically less stable than seed=0 with this cosine schedule:

1. **Epochs 1–96**: Steady improvement, best val 11.888% at ep96
2. **Epoch 97**: Catastrophic spike — val jumps from 11.888% → 62.8%, aligning with cosine LR restart boundary (T_max=30, so restarts at ep91)
3. **Epochs 100–150**: Partial recovery to 12.9%, but never matched ep96 quality
4. **Epochs 150–387**: Progressive deterioration with increasing instability. Val oscillated: ep200 (19.8%) → ep250 (55.3%) → ep350 (60.9%)
5. **Epochs 370–387**: Gradient norms exploded to 10^16 (e.g. ep373: 2.5×10^16)
6. **Epoch 388+**: Full NaN — training completely diverged. Terminal model is NaN.

The full test eval on the best checkpoint (ep96) yielded **12.155%** — this is the paper-facing number but it's 3× worse than the seed=0 baseline (3.997%) because the model only had 96 useful epochs before diverging.

### Suggested follow-ups

- **Gradient clipping**: `--grad-clip 1.0` would likely prevent the NaN divergence and allow seed=123 to survive the cosine restarts, enabling convergence past ep96
- **Seed variance**: The 3× gap between seed=0 (3.997%) and seed=123 (11.888%) suggests high seed sensitivity for this DrivAerML config — this is important context for multi-seed paper-facing results
- Should I retry with `--grad-clip 1.0` added to the same seed=123 command?

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3064 because ADVISOR: Multi-seed s123 on old config (no EMA+gc) — 11.888%, diverged ep388. Same issue as #3065 (s456). Not valid paper-facing data. Multi-seed runs should wait for post-noam-pivot best config. Reassigning to noam-feature experiment..
